test/test_overlapping_loc.ml

```diff
@@ -0,0 +1,43 @@
+let test_1 () =
+  (* [cas_2] acts on the same location as [cas_1]
+     and has conflicting values. kCAS should fail.
+  *)
+  let v_1 = Kcas.ref 0 in
+  let cas_1 = Kcas.mk_cas v_1 0 1 in
+  let cas_2 = Kcas.mk_cas v_1 2 3 in
+
+  match Kcas.kCAS [ cas_1; cas_2 ] with exception _ -> () | _ -> assert false
```

Comment:
This test (and `test_3` below) makes an assumption that the cas ops in the list given to kCAS are performed sequentially in the list order.  If the algorithm decides to try the `cas_2` first, the whole kCAS fails and returns false before detecting overlap.  I believe it would be best to avoid that assumption and allow the algorithm to choose whichever order it wants.  One way to avoid the assumption would be to run kCAS twice with both orders (`Kcas.kCAS [ cas_1; cas_2 ] || Kcas.kCAS [ cas_2; cas_1 ]`) and assert that one of them raises an exception (the other might just return false).

Reply:
Yeah, that's not ideal; moving the check to the beginning of kCAS function should fix it.